File: src/main/kotlin/at/fhhgb/mc/component/sim/model/SimulationModel.java
package at.fhhgb.mc.component.sim.model;

import at.fhhgb.mc.component.base.BaseModel;
import at.fhhgb.mc.component.sim.domain.FloorManager;
import at.fhhgb.mc.component.sim.domain.PixelPosition;
import at.fhhgb.mc.component.sim.model.initialisations.WaypointPosition;
import at.fhhgb.mc.component.sim.model.initialisations.WaypointRoute;
import at.fhhgb.mc.component.sim.model.recorder.network.AccessPoint;
import at.fhhgb.mc.component.sim.model.simulator.Floor;
import at.fhhgb.mc.component.sim.positioning.IPositioning;
import at.fhhgb.mc.component.sim.sectorizing.ISectorizing;
import at.fhhgb.mc.component.sim.view.DrawingContext;

import java.awt.*;
import java.awt.image.BufferedImage;
import java.util.ArrayList;
import java.util.List;

/**
 * The Model of the MVC pattern. The new positions and polygons are drawn here.
 * 
 * @author Michael Nigl
 */
public class SimulationModel extends BaseModel {

	List<AccessPoint> accessPoints = new ArrayList<>();
	Person person = new Person(-1000,1000);
    Person actualPosition = new Person(-1000,1000);
	WaypointRoute waypointRoute;
	private int m_mode = 4;
	private int xGrid = 1;
    private int yGrid = 1;
    private int[] wayPointCount;
    private int wayPointNumber = 1;
    private int interpolationStep = 0;
	private ISectorizing sectorizing;

	private IPositioning positioning;


	/**
	 * Empty default constructor
	 */
	public SimulationModel() {
	}

	public void generateFloorMap(Floor floor, int[] dimensions) {
		setPanelSize(dimensions[0], dimensions[1]);
		accessPoints = floor.getAccessPointList();
		if (floorManager == null) {
			floorManager = new FloorManager(floor);
		} else {
			floorManager.changeFloor(floor);
		}
		zoomToFit(true);
	}

    private void checkForinitializedWaypoints(){
        if(waypointRoute == null){
            waypointRoute = new WaypointRoute();
        }
        Position pos = null;
        if(wayPointNumber < wayPointCount.length){
            pos = interpolate(waypointRoute.getWayPointPositionAtIndex(wayPointNumber-1).getWayPointPosition(),waypointRoute.getWayPointPositionAtIndex
                    (wayPointNumber).getWayPointPosition(),wayPointCount[wayPointNumber-1],interpolationStep);
        }else{
            pos = interpolate(waypointRoute.getWayPointPositionAtIndex(wayPointNumber-1).getWayPointPosition(),waypointRoute.getWayPointPositionAtIndex
                    (0).getWayPointPosition(),wayPointCount[wayPointNumber-1],interpolationStep);
        }

        actualPosition.setX(pos.getX());
        actualPosition.setY(pos.getY());

        interpolationStep+=1;
        if(interpolationStep == wayPointCount[wayPointNumber-1]){
            wayPointNumber+=1;
            interpolationStep = 0;
        }
    }


    public void setWayPointCount(int[] wayPointCount){
        this.wayPointCount = wayPointCount;
    }

    public void calculateSimplePersonPosition(List<AccessPoint> accessPoints){
        checkForinitializedWaypoints();
        for(AccessPoint accessPoint : accessPoints){
            for(AccessPoint ap : this.accessPoints){
                if(ap.getBSSID().toLowerCase().equals(accessPoint.getBSSID().toLowerCase())){
                    person.setX(ap.getPosition().getX());
                    person.setY(ap.getPosition().getY());
                    callObserver(createBufferedImage());
                    return;
                }
            }
        }

		callObserver(createBufferedImage());
	}

    public void calculateWeightedPersonPosition(List<AccessPoint> accessPointList){
        checkForinitializedWaypoints();
        int offset = 12;
        List<AccessPoint> matchingAccesspoints = new ArrayList<>();
        for(AccessPoint accessPoint : accessPointList){
            for(AccessPoint ap : accessPoints){
                if(ap.getBSSID().toLowerCase().equals(accessPoint.getBSSID().toLowerCase())){
                    ap.setSignalLevel(accessPoint.getSignalLevel());
                    matchingAccesspoints.add(ap);
                }
            }
        }

        double weightsum = 0;
        double positionSumX = 0;
        double positionSumY = 0;

        for(AccessPoint accessPoint : matchingAccesspoints){

            positionSumX += (accessPoint.getPosition().getX() * (accessPoint.getSignalLevel()*-1));
            positionSumY += (accessPoint.getPosition().getY() * (accessPoint.getSignalLevel()*-1));

            weightsum += (accessPoint.getSignalLevel()*-1);
        }

        person.setX((int)(positionSumX/weightsum));
        person.setY((int)(positionSumY/weightsum));
		callObserver(createBufferedImage());
    }


    public Position interpolate(Position start, Position end, int steps, int iterationStep) {
        if (steps < 2) {
			return new Position(person.getX(),person.getY());
            //throw new IllegalArgumentException("interpolate: illegal count!");
        }

        int xPos = (int)Math.round(start.getX() + iterationStep * (end.getX() - start.getX()) / steps);
        int yPos = (int)Math.round(start.getY() + iterationStep * (end.getY() - start.getY()) / steps);


        return new Position(xPos,yPos);
    }
	
	/**
	 * Calculates for a point on the screen the values it has in the world coordinate system.
	 * 
	 * @param _pt - the point on the screen
	 * @return - the point in the world coordinate system
	 */
	public Point getMapPoint(Point _pt){
		Point search = new Point(_pt.x,_pt.y);
		search = m_tMatrix.invers().multiply(search);
		return search;
	}


	public void changeXGrid(int gridWitdh){
		this.xGrid = gridWitdh;
		callObserver(createBufferedImage());
	}

	public void changeYGrid(int gridHeight){
		this.yGrid = gridHeight;
		callObserver(createBufferedImage());
	}

	public void setSectorizing(ISectorizing sectorizing) {
		this.sectorizing = sectorizing;
		callObserver(createBufferedImage());
	}

	public IPositioning getPositioning() {
		return positioning;
	}

	public void setPositioning(IPositioning positioning) {
		this.positioning = positioning;
	}

	public void addPlayerData(List<AccessPoint> accessPointList) {
		checkForinitializedWaypoints();
		Position p = positioning.calculatePosition(accessPointList);
		if (sectorizing == null || p == null) {
			callObserver(createBufferedImage());
			return;
		}
		PixelPosition newPos = floorManager.calculatePixelPositionFromMeter(p.getX(), p.getY());
		sectorizing.addCurrentPosition(new Position(Math.round(newPos.getXPosition()), Math.round(newPos.getYPosition())));
		callObserver(createBufferedImage());
	}

	/**
	 * Creates a buffered image of all the current GeoObjects. It is used because it is faster
	 * than the usage of single objects.
	 *
	 * @return the created buffered image.
	 */
	public BufferedImage createBufferedImage(){
		BufferedImage images = applyMatrixToBufferedImage(floorManager.getFloor().getFloorImage());
		Graphics g = images.createGraphics();
		Rectangle clipRect = new Rectangle(floorManager.getOffsetLeftInPixel() - 10, floorManager.getOffsetTopInPixel() - 10, floorManager.getOffsetRightInPixel() - floorManager.getOffsetLeftInPixel() + 15, floorManager.getOffsetBottomInPixel() - floorManager.getOffsetTopInPixel() + 15);
		g.setClip(new Rectangle(m_tMatrix.multiply(clipRect)));
		g.setColor(new Color(0, 0, 0));

		g.drawOval(0, 0, 10, 10);
		int i = 0;

		i = 0;
		while(i < accessPoints.size()) {
			PixelPosition pos = floorManager.calculatePixelPositionFromMeter(accessPoints.get(i).getPosition().getX(), accessPoints.get(i).getPosition().getY());
			DrawingContext.drawAccessPoint(pos.getXPosition(), pos.getYPosition(), g, m_tMatrix);
			i++;
		}

		if(waypointRoute != null) {
			i = 0;
			List<WaypointPosition> wayPointPositions = waypointRoute.getWaypointPositionList();
			List<WaypointPosition> wayPointPositions2 = new ArrayList<>();

			while (i < wayPointPositions.size()) {
				WaypointPosition waypointPosition = wayPointPositions.get(i);
				PixelPosition pos = floorManager.calculatePixelPositionFromMeter(waypointPosition.getWayPointPosition().getX(), waypointPosition.getWayPointPosition().getY());
				WaypointPosition waypointPosition1 = new WaypointPosition(waypointPosition.getWayPointId(),
						new Position(pos.getXPosition(), pos.getYPosition()));
				wayPointPositions2.add(waypointPosition1);

				DrawingContext.drawWayPointPosition(waypointPosition1, g, m_tMatrix);
				i++;
			}
			DrawingContext.drawWayPointLine((ArrayList<WaypointPosition>) wayPointPositions2, g, m_tMatrix);
		}

		if (sectorizing != null) {
			sectorizing.createSectors(floorManager, g, m_tMatrix);
		}

		PixelPosition pos = floorManager.calculatePixelPositionFromMeter(person.getX(), person.getY());
		person.setX(pos.getXPosition());
		person.setY(pos.getYPosition());
		DrawingContext.drawPerson(person, g, m_tMatrix);

		pos = floorManager.calculatePixelPositionFromMeter(actualPosition.getX(), actualPosition.getY());
		actualPosition.setX(pos.getXPosition());
		actualPosition.setY(pos.getYPosition());
		DrawingContext.drawActualPosition(actualPosition,g,m_tMatrix);


		return images;
	}




}

File: src/main/kotlin/at/fhooe/mc/wifipositioning/model/simulation/SimulationModel.kt
package at.fhooe.mc.wifipositioning.model.simulation

import at.fhooe.mc.wifipositioning.interfaces.PlaybackCallbackInterface
import at.fhooe.mc.wifipositioning.model.configuration.ConfigurationModel
import at.fhooe.mc.wifipositioning.model.graphics.FloorManager
import at.fhooe.mc.wifipositioning.model.simulation.recording.ScannedAccessPoint
import at.fhooe.mc.wifipositioning.model.simulation.simulator.Floor
import at.fhooe.mc.wifipositioning.model.positioning.IPositioning
import at.fhooe.mc.wifipositioning.model.sectoring.ISectoring
import at.fhooe.mc.wifipositioning.model.graphics.DrawingContext
import at.fhooe.mc.wifipositioning.model.building.InstalledAccessPoint
import at.fhooe.mc.wifipositioning.model.initialisations.Route
import at.fhooe.mc.wifipositioning.model.initialisations.Waypoint
import at.fhooe.mc.wifipositioning.utility.Player

import java.awt.*
import java.awt.image.BufferedImage
import java.util.*

/**
 * The Model of the MVC pattern. The new positions and polygons are drawn here.
 */
class SimulationModel(var config: ConfigurationModel) : BaseModel(), PlaybackCallbackInterface, Observer {
    private var accessPoints: List<InstalledAccessPoint>? = ArrayList()
    private var person = Position(-1000, 1000)
    private var actualPosition = Position(-1000, 1000)

    private var route: Route? = null
        get() = config.route

    private var wayPointCount: IntArray? = null
    private var wayPointNumber = 1
    private var interpolationStep = 0

    private var player: Player? = null
    private var playerThread: Thread? = null

    val sectoring: ISectoring?
        get() = config.sectoring

    val positioning: IPositioning?
        get() = config.positioning

    init {
        config.addObserver(this)
    }

    fun reloadConfiguration() {
        config.loadFloors()

        config.building?.let { building ->
            generateFloorMap(building.getFloor(3))
        }

        config.loadBuildingGraph()
        config.loadPositioningMethod()
        config.loadWaypoints()
        player = config.loadWalkRecording(this)
    }

    fun toggleSimulation() {
        if (player?.isRunning == true) {
            player?.pausePlayback = player?.pausePlayback != true
            return
        }

        player?.let { player ->
            playerThread = (object : Thread() {
                override fun run() {
                    player.startPlayback()
                }
            })

            playerThread?.let {
                it.start()
            }
        }
    }

    fun resetSimulation() {
        player?.stopPlayback = true
        playerThread = null

        person = Position(-1000, 1000)
        actualPosition = Position(-1000, 1000)

        wayPointNumber = 1
        interpolationStep = 0

        config?.resetPositioning()
        config?.resetSectoring()
    }

    override fun generateFloorMap(floor: Floor) {
        accessPoints = floor.accessPointList
        if (floorManager == null) {
            floorManager = FloorManager(floor)
        } else {
            floorManager?.changeFloor(floor)
        }
    }

    private fun checkForInitializedWaypoints() {
        if(player?.isRunning == false || player?.stopPlayback == true || route == null) {
            return
        }

        var pos: Position? = null
        route?.let { route ->
            if (wayPointNumber < wayPointCount!!.size) {
                pos = interpolate(route.waypointList[wayPointNumber - 1].position, route.waypointList[wayPointNumber].position, wayPointCount!![wayPointNumber - 1], interpolationStep)
            } else {
                wayPointCount?.let {
                    pos = interpolate(route.waypointList[wayPointNumber - 1].position, route.waypointList[0].position, wayPointCount!![wayPointNumber - 1], interpolationStep)
                }
            }
        }

        pos?.let {pos ->
            actualPosition.x = pos.x
            actualPosition.y = pos.y

            interpolationStep += 1
            if (interpolationStep == wayPointCount!![wayPointNumber - 1]) {
                wayPointNumber += 1
                interpolationStep = 0
            }
        }
    }

    private fun calculateSimplePersonPosition(scannedAccessPoints: List<ScannedAccessPoint>) {
        checkForInitializedWaypoints()
        for (accessPoint in scannedAccessPoints) {
            for (ap in this.accessPoints!!) {
                if (ap.bssid.toLowerCase() == accessPoint.bssid.toLowerCase()) {
                    person.x = ap.position.x
                    person.y = ap.position.y
                    callObserver(createBufferedImage() ?: return)
                    return
                }
            }
        }

        callObserver(createBufferedImage() ?: return)
    }

    private fun interpolate(start: Position, end: Position, steps: Int, iterationStep: Int): Position {
        if (steps < 2) {
            return Position(person.x, person.y)
        }

        val xPos = Math.round((start.x + iterationStep * (end.x - start.x) / steps).toFloat())
        val yPos = Math.round((start.y + iterationStep * (end.y - start.y) / steps).toFloat())

        return Position(xPos, yPos)
    }

    private fun addPlayerData(scannedAccessPointList: List<ScannedAccessPoint>) {
        checkForInitializedWaypoints()
        val p = positioning?.calculatePosition(scannedAccessPointList)
        if (sectoring == null || p == null) {
            callObserver(createBufferedImage() ?: return)
            return
        }
        val newPos = floorManager?.calculatePixelPositionFromMeter(p.position.x, p.position.y)
        newPos?.let {
            sectoring!!.addCurrentPosition(Position(Math.round(newPos.x.toFloat()), Math.round(newPos.y.toFloat())))
            callObserver(createBufferedImage() ?: return)
        }
    }

    /**
     * Creates a buffered image of all the current GeoObjects. It is used because it is faster
     * than the usage of single objects.
     *
     * @return the created buffered image.
     */
    override fun createBufferedImage(): BufferedImage? {
        val floorManager = this.floorManager ?: return null

        val images = applyMatrixToBufferedImage(floorManager.floor?.floorImage ?: return null)

        val g = images.createGraphics()
        val clipRect = Rectangle(floorManager.offsetLeftInPixel - 10, floorManager.offsetTopInPixel - 10, floorManager.offsetRightInPixel - floorManager.offsetLeftInPixel + 15, floorManager.offsetBottomInPixel - floorManager.offsetTopInPixel + 15)
        g.clip = Rectangle(matrix.multiply(clipRect))
        g.color = Color(0, 0, 0)

        g.drawOval(0, 0, 10, 10)
        var i = 0

        i = 0

        accessPoints?.let { accessPoints ->
            while (i < accessPoints.size) {
                val ap = accessPoints[i]
                val pos = floorManager.calculatePixelPositionFromMeter(ap.position.x, ap.position.y)
                DrawingContext.drawAccessPoint(ap, pos.x, pos.y, g, matrix)
                i++
            }
        }

        if (route != null) {
            i = 0
            val wayPointPositions = route!!.waypointList
            val wayPointPositions2 = ArrayList<Waypoint>()

            while (i < wayPointPositions.size) {
                val waypointPosition = wayPointPositions[i]
                val pos = floorManager.calculatePixelPositionFromMeter(waypointPosition.position.x, waypointPosition.position.y)
                val waypointPosition1 = Waypoint(waypointPosition.id, waypointPosition.description, waypointPosition.floor, Position(pos.x, pos.y))
                wayPointPositions2.add(waypointPosition1)

                DrawingContext.drawWayPointPosition(waypointPosition1, g, matrix)
                i++
            }
            DrawingContext.drawWayPointLine(wayPointPositions2, g, matrix)
        }

        if (sectoring != null) {
            sectoring!!.createSectors(floorManager, g, matrix)
        }

        var pos = floorManager.calculatePixelPositionFromMeter(person.x, person.y)
        person.x = pos.x
        person.y = pos.y
        DrawingContext.drawPerson(person, g, matrix)

        pos = floorManager.calculatePixelPositionFromMeter(actualPosition.x, actualPosition.y)
        actualPosition.x = pos.x
        actualPosition.y = pos.y
        DrawingContext.drawActualPosition(actualPosition, g, matrix)


        return images
    }

    override fun nearestAccessPoint(scannedAccessPoints: MutableList<ScannedAccessPoint>?) {
        scannedAccessPoints?.let {
            calculateSimplePersonPosition(scannedAccessPoints)
        }
    }

    override fun allAccessPoints(accessPointList: MutableList<ScannedAccessPoint>?) {
        accessPointList?.let {
            addPlayerData(accessPointList)
        }
    }

    override fun wayPointCount(wayPointCount: IntArray?) {
        wayPointCount?.let {
            this.wayPointCount = wayPointCount
        }
    }

    override fun update(o: Observable?, arg: Any?) {
        if(o != config) return

        resetSimulation()
        reloadConfiguration()

        println("Reloaded configuration.")
    }
}
